src/v/compat/run.cc

```diff
@@ -0,0 +1,196 @@
+/*
+ * Copyright 2022 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#include "compat/run.h"
+
+#include "compat/check.h"
+#include "compat/metadata_dissemination_compat.h"
+#include "compat/raft_compat.h"
+#include "json/document.h"
+#include "json/prettywriter.h"
+#include "json/writer.h"
+#include "seastarx.h"
+#include "utils/base64.h"
+#include "utils/file_io.h"
+
+#include <seastar/core/thread.hh>
+
+namespace compat {
+
+template<typename... Types>
+struct type_list {};
+
+using compat_checks = type_list<
+  raft::timeout_now_request,
+  raft::timeout_now_reply,
+  raft::transfer_leadership_request,
+  raft::transfer_leadership_reply,
+  cluster::update_leadership_request>;
+
+struct compat_error final : public std::runtime_error {
+public:
+    explicit compat_error(std::string_view msg)
+      : std::runtime_error(fmt::format("compat: {}", msg)) {}
+};
+
+template<typename T>
+struct corpus_helper {
+    using checker = compat_check<T>;
+
+    /*
+     * Builds a test case for an instance of T.
+     *
+     * {
+     *   "name": "raft::some_request",
+     *   "fields": { test case field values },
+     *   "binaries": [
+     *     {
+     *       "name": "serde",
+     *       "data": "base64 encoding",
+     *     },
+     *   ]
+     * }
+     */
+    static void write_test_case(T t, json::Writer<json::StringBuffer>& w) {
+        auto&& [ta, tb] = compat_copy(std::move(t));
+
+        w.StartObject();
+        w.Key("name");
+        w.String(checker::name.data());
+
+        w.Key("fields");
+        w.StartObject();
+        checker::to_json(std::move(ta), w);
+        w.EndObject();
+
+        w.Key("binaries");
+        w.StartArray();
+        for (auto& b : checker::to_binary(std::move(tb))) {
+            w.StartObject();
+            w.Key("name");
+            w.String(b.name);
+            w.Key("data");
+            w.String(iobuf_to_base64(b.data));
+            w.EndObject();
+        }
+        w.EndArray();
+        w.EndObject();
+    }
+
+    /*
+     * Writes all test cases to the provided directory.
+     */
+    static ss::future<> write(const std::filesystem::path& dir) {
+        size_t instance = 0;
+
+        for (auto& test : checker::create_test_cases()) {
+            // json encoded test case
+            auto buf = json::StringBuffer{};
+            auto writer = json::Writer<json::StringBuffer>{buf};
+            write_test_case(std::move(test), writer);
+
+            // save test case to file
+            iobuf data;
+            data.append(buf.GetString(), buf.GetSize());
+            auto fn = fmt::format("{}_{}.json", checker::name, instance++);
+            write_fully(dir / fn, std::move(data)).get();
```

Comment:
If we are going with `ss::future<>` as the return type, why not write this method without `.get()` ? i.e. `co_await write_fully(...)`

Reply:
seastar disallows co_await/coro in a ss::async context, and it's caller, write_context is in ss::async context because clang has a compiler bug in 14 that prevents coroutines from working reliably with templates (see bug references about representations across sections).

Reply:
I didn't mean use `coroutines` i was just writing that as an example, you could just `ss::do_for_each({ ... return write_fully() });`

Reply:
Its fine, just upon first glance seems strange